### ISSUE TITLE
Resnet TCN version

### ISSUE THREAD

**Issue by gitproxy** [abedicodes/ResNet-TCN#3]
1.  I noticed that the dataloader class has issues loading batches that are mismatched sizes (batch size 5 leaves 4 samples remainder for the test set) How do you handle this?
2. May I ask if this source code is the version in your paper with the sampling and weighted loss implemented (ResNet + TCN with weighted sampling and weighted loss). 
3. However the version in your paper with the highest test accuracy does not have these features implemented (the one with 63.9 % accuracy) What modifications to the code is needed to achieve the 63.9 % accuracy?

**Comment by abediai** [comment]
To solve the issue of the remainder of non-full batches (e.g., of size 4), set drop_last=True in the DataLoader in utils.generate_dataloader.

To achieve the 63.9 accuracy, just use weighted sampling in the DataLoader (without using a weighted loss function),

```
def make_weights_for_balanced_classes(videos, nclasses):
    count = [0] * nclasses
    for item in videos:
        count[item[1]] += 1
    weight_per_class = [0.] * nclasses
    N = float(sum(count))
    for i in range(nclasses):
        weight_per_class[i] = N / float(count[i])
    weight = [0] * len(videos)
    for idx, val in enumerate(videos):
        weight[idx] = weight_per_class[val[1]]
    return weight


weights = make_weights_for_balanced_classes(dataset.imgs, len(dataset.classes))
weights = torch.DoubleTensor(weights)
sampler = torch.utils.data.sampler.WeightedRandomSampler(weights, len(weights))
dataloader = torch.utils.data.DataLoader(dataset, batch_size=batch_size, sampler=sampler, drop_last=True, num_workers=4)

```

**Comment by gitproxy** [comment]
Alright thank you

**Comment by gitproxy** [comment]
Hi, apologies for going back to this query, but I was wondering where the code is supposed to go? Is it supposed to be put into utils.py? Furthermore, where should the code be modified?

**Comment by abediai** [comment]
Yes, first, using Pandas, extract the labels of the samples in four classes from the CSV file in the DAiSEE dataset, and then convert them to a NumPy array. The NumPy array should be the input to the above function. Add the above function to utils.py.

**Comment by patnaik42** [comment]
Hi,

In addition to above mentioned weighted sampling, I have tried with the below parameters as mentioned in the paper to get 63.9 accuracy with DAiSEE dataset. But training itself is not converging and training accuracy doesn't go beyond  0.5 even till 100 epochs. Can you please confirm if the below parameters are correct to get 63.9 accuracy? If not, can you provide correct parameters to be used?


L = 50  #  Number of frames at down sampled rate
Batch-size = 5

ResTCN module parameters:

 self.spatial_feat_dim = 32
 self.num_classes = 4
 self.nhid = 128
 self.levels = 8
 self.kernel_size = 7
 self.dropout = .25

**Comment by patnaik42** [comment]
> Hi,
> 
> In addition to above mentioned weighted sampling, I have tried with the below parameters as mentioned in the paper to get 63.9 accuracy with DAiSEE dataset. But training itself is not converging and training accuracy doesn't go beyond 0.5 even till 100 epochs. Can you please confirm if the below parameters are correct to get 63.9 accuracy? If not, can you provide correct parameters to be used?
> 
> L = 50 # Number of frames at down sampled rate Batch-size = 5
> 
> ResTCN module parameters:
> 
> self.spatial_feat_dim = 32 self.num_classes = 4 self.nhid = 128 self.levels = 8 self.kernel_size = 7 self.dropout = .25


As mentioned in the paper,  I have also tried self.spatial_feat_dim = 512 but I still see the same issue

**Comment by qq1332427275** [comment]
> Hi, apologies for going back to this query, but I was wondering where the code is supposed to go? Is it supposed to be put into utils.py? Furthermore, where should the code be modified?

Hi, have you successfully reproduced the 63.9% accuracy？

**Comment by qq1332427275** [comment]
@gitproxy Hi, have you successfully reproduced the 63.9% accuracy ？
